### PR TITLE
Make overlay home dir Private mount

### PR DESCRIPTION
**Pull request by rhvgoyal**
People have reported following issue with overlay

$ docker run -ti --name=foo -v /dev/:/dev fedora bash
$ docker cp foo:/bin/bash /tmp
$ exit container

Upon container exit, /dev/pts gets unmounted too. This happens because
docker cp volume mounts get propagated to /run/docker/libcontainer/....
and when container exits, it must be tearing down mount point under
/run/docker/libcontainerd/... and as these are "shared" mounts it
propagates events to /dev/pts and it gets unmounted too.

One way to solve this problem is to make sure "docker cp" volume mounts
don't become visible under /run/docker/libcontainerd/..

Here are more details of what is actually happening.

Make overlay home directory (/var/lib/docker/overlay) private mount when
docker starts and unmount it when docker stops. Following is the reason
to do it.

In fedora and some other distributions / is "shared". That means when
docker creates a container and mounts it root in /var/lib/docker/overlay/...
that mount point is "shared".

Looks like after that containerd/runc bind mounts that rootfs into
/runc/docker/libcontainerd/container-id/rootfs. And this puts both source
and destination mounts points in shared group and they both are setup
to propagate mount events to each other.

Later when "docker cp" is run it sets up container volumes under
/var/lib/dokcer/overlay/container-id/... And all these mounts propagate
to /runc/docker/libcontainerd/... Now mountVolumes() makes these new
mount points private but by that time propagation already has happened
and private only takes affect when unmount happens.

So to stop this propagation of volumes by docker cp, make
/var/lib/docker/overlay a private mount point. That means when a container
rootfs is created, that mount point will be private too (it will inherit
property from parent). And that means when bind mount happens in /runc/
dir, overlay mount point will not propagate mounts to /runc/.

Other graphdrivers like devicemapper are already doing it and they don't
face this issue.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

### CODE REVIEW

**Comment by rhvgoyal** [comment]
cc @rhatdan @cpuguy83 @thaJeztah @AkihiroSuda

**Comment by rhatdan** [comment]
@runcom We need to back port this.

File: daemon/graphdriver/overlay/overlay.go [hunk 1]
```diff
@@ -220,6 +225,10 @@ func (d *Driver) GetMetadata(id string) (map[string]string, error) {
 // Cleanup simply returns nil and do not change the existing filesystem.
 // This is required to satisfy the graphdriver.Driver interface.
 func (d *Driver) Cleanup() error {
+	if err := mount.Unmount(d.home); err != nil {
+		return err
+	}
+
 	return nil
```
Comment (rhatdan) [inline daemon/graphdriver/overlay/overlay.go]: Could have just done 

return mount.Unmount(d.home)

Comment (rhvgoyal) [inline daemon/graphdriver/overlay/overlay.go]: Right. That's simpler. Will change it.

**Comment by AkihiroSuda** [comment]
@rhvgoyal Thank you for working on this. LGTM (non-binding).
I tested on Ubuntu 15.10.

**Comment by rhvgoyal** [comment]
ping @cpuguy83

**Comment by cpuguy83** [comment]
LGTM

File: daemon/graphdriver/overlay/overlay.go [hunk 2]
```diff
@@ -220,7 +225,7 @@ func (d *Driver) GetMetadata(id string) (map[string]string, error) {
 // Cleanup simply returns nil and do not change the existing filesystem.
```
Comment (LK4D4) [inline daemon/graphdriver/overlay/overlay.go]: comment is not true anymore

Comment (rhvgoyal) [inline daemon/graphdriver/overlay/overlay.go]: Right. will fix it.

**Comment by rhvgoyal** [comment]
@LK4D4 Fixed the comment and pushed again.

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**